File: graph/Programmers/[Programmers] TargetNumber.go
type Queue []interface{}

func (q *Queue) IsEmpty() bool {
	return len(*q) == 0
}

func (q *Queue) Push(v interface{}) {
	*q = append(*q, v)
}

func (q *Queue) Pop() (v interface{}) {
	if q.IsEmpty() {
		return nil
	}

	v = (*q)[0]
	*q = (*q)[1:]
	return
}

type Node struct {
	Now   int
	Value int
}

func solution(numbers []int, target int) (result int) {
	op := []int{-1, 1}
	length := len(numbers) - 1
	q := Queue{}

	q.Push(Node{0, numbers[0]})
	q.Push(Node{0, numbers[0] * -1})

	for !q.IsEmpty() {
		v := q.Pop().(Node)

		if v.Now == length-1 {
			r1 := v.Value + op[0]*numbers[v.Now+1]
			r2 := v.Value + op[1]*numbers[v.Now+1]

			if r1 == target {
				result++
			}
			if r2 == target {
				result++
			}
			continue
		}

		q.Push(Node{v.Now + 1, v.Value + op[0]*numbers[v.Now+1]})
		q.Push(Node{v.Now + 1, v.Value + op[1]*numbers[v.Now+1]})
	}

	return
}

File: graph/Programmers/[Programmers] TargetNumber.cpp
//프로그래머스
//BFS - 타겟 넘버
#include <string>
#include <vector>
#include <utility>
#include <queue>

using namespace std;

int operation[2] = {1, -1};

int solution(vector<int> numbers, int target) {
    int answer = 0, idx = 1;
    int size = numbers.size();
    queue<pair<int,int>> q;
    q.push(make_pair(numbers[0],0));
    q.push(make_pair(-1 * numbers[0],0));
    
    while(!q.empty()) {
        int now = q.front().first, idx = q.front().second;
        q.pop();
        
        if(idx == size - 1) {
            if(now == target) {
                answer++;     
            }
            continue;
        }
        
        for(int opIdx = 0; opIdx < 2 && idx < size - 1; opIdx++) {
            q.push(make_pair(now + (numbers[idx + 1] * operation[opIdx]),idx + 1));
        }
    }
    
    return answer;
}